File: typescript/src/n0733_flood_fill.ts
export function floodFill(
  image: number[][],
  startRow: number,
  startCol: number,
  fillColor: number
): number[][] {
  const startColor = image[startRow]![startCol];
  if (startColor === fillColor) {
    return image;
  }
  image[startRow]![startCol] = fillColor;
  const stack: [number, number][] = [[startRow, startCol]];
  while (stack.length > 0) {
    const [row, col] = stack.pop()!;
    const adjacent: [number, number][] = [
      [row - 1, col],
      [row, col - 1],
      [row + 1, col],
      [row, col + 1],
    ];
    for (const [r, c] of adjacent) {
      if (
        0 <= r &&
        r < image.length &&
        0 <= c &&
        c < image[0]!.length &&
        image[r]![c] === startColor
      ) {
        image[r]![c] = fillColor;
        stack.push([r, c]);
      }
    }
  }
  return image;
}

File: python/src/n0733_flood_fill.py
def flood_fill(
    image: list[list[int]], start_row: int, start_col: int, fill_color: int
) -> list[list[int]]:
    start_color = image[start_row][start_col]
    if start_color == fill_color:
        return image
    image[start_row][start_col] = fill_color
    stack = [(start_row, start_col)]
    while stack:
        (row, col) = stack.pop()
        adjacent = [(row - 1, col), (row, col - 1), (row + 1, col), (row, col + 1)]
        for r, c in adjacent:
            if (
                (0 <= r < len(image))
                and (0 <= c < len(image[0]))
                and image[r][c] == start_color
            ):
                image[r][c] = fill_color
                stack.append((r, c))
    return image
